contracts/SafeProtocolMediator.sol

```diff
@@ -0,0 +1,241 @@
+    error RootAccessActionExecutionFailed(address safe, bytes32 metaHash);
+    error ModuleAlreadyEnabled(address safe, address module);
+    error InvalidModuleAddress(address module);
+    error InvalidPrevModuleAddress(address module);
+    error ZeroPageSizeNotAllowed();
+
+    modifier onlyEnabledModule(ISafe safe) {
+        if (enabledModules[address(safe)][msg.sender].nextModulePointer == address(0)) {
+            revert MoudleNotEnabled(msg.sender);
+        }
+        _;
+    }
+
+    modifier noZeroOrSentinelModule(address module) {
+        if (module == address(0) || module == SENTINEL_MODULES) {
+            revert InvalidModuleAddress(module);
+        }
+        _;
+    }
+
+    /**
+     * @notice This function executes a delegate call on a safe if the module is enabled and
+     *         root access it granted.
+     * @param safe A Safe instance
+     * @param transaction A struct of type SafeTransaction containing information of about the action(s) to be executed.
+     *                    Users can add logic to validate metahash through a transaction guard.
+     * @return data bytes types containing the result of the executed action.
+     */
+    function executeTransaction(
+        ISafe safe,
+        SafeTransaction calldata transaction
+    ) external override onlyEnabledModule(safe) returns (bytes[] memory data) {
+        data = new bytes[](transaction.actions.length);
+        for (uint256 i = 0; i < transaction.actions.length; ++i) {
+            SafeProtocolAction memory safeProtocolAction = transaction.actions[i];
+            (bool isActionSuccessful, bytes memory resultData) = safe.execTransactionFromModuleReturnData(
+                safeProtocolAction.to,
+                safeProtocolAction.value,
+                safeProtocolAction.data,
+                0
+            );
+
+            // Need to revisit the approach below. If some actions fail, the transaction stiil succeeds.
+            // With current approach, even if one action fails, `data` will be empty bytes even for successful
+            // actions.
+            if (!isActionSuccessful) {
+                revert ActionExecutionFailed(address(safe), transaction.metaHash, i);
+            } else {
+                data[i] = resultData;
+            }
+        }
+
+        emit ActionsExecuted(address(safe), transaction.metaHash, transaction.nonce);
+    }
+
+    /**
+     * @notice This function executes a delegate call on a safe if the module is enabled and
+     *         root access it granted.
+     * @param safe A Safe instance
+     * @param rootAccess A struct of type SafeRootAccess containing information of about the action to be executed.
+     *                   Users can add logic to validate metahash through a transaction guard.
+     * @return data bytes types containing the result of the executed action.
+     */
+    function executeRootAccess(
+        ISafe safe,
+        SafeRootAccess calldata rootAccess
```

Comment:
The difference between this struct and the other one is minimal. Basically, it only allows 1 transaction instead of multiple; is there a specific reason for this?

Reply:
This had been moved to https://github.com/5afe/safe-protocol-specs/discussions/6 to add proper documentation